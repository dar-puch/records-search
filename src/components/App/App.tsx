import * as React from "react";
import { useState } from "react";
import { RecordsList } from "../RecordsList";
import { what, Release } from "../../utils/api";
import { ErrorInfo } from "../ErrorInfo";
import { QueryHistory } from "../QueryHistory";
import {
  Button,
  Input,
  AppBar,
  Toolbar,
  Typography,
  CircularProgress,
  Box,
} from "@material-ui/core";
import { makeStyles } from "@material-ui/core/styles";
import { fetchArtist, useQueryHistory } from "../../utils/functions";

export enum responseStateValues {
  "idle",
  "loading",
  "success",
  "failure",
}

const useStyles = makeStyles((theme) => ({
  toolbar: {
    minHeight: 128,
    alignItems: "flex-start",
    paddingTop: theme.spacing(1),
    paddingBottom: theme.spacing(2),
  },
  button: {
    margin: 20,
  },
}));

const App = (): JSX.Element => {
  const [recList, setRecList] = useState<Release[]>([]);
  const [inputValue, setInputValue] = useState("");
  const [error, setError] = useState("");
  const [responseState, setResponseState] = useState<responseStateValues>(
    responseStateValues.idle
  );
  const {
    saveQuery,
    deleteHistoryItem,
    clearHistory,
    queryHistory,
  } = useQueryHistory();

  const classes = useStyles("root");

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setInputValue(e.target.value);
  };

  const saveSearchResult = async (what: what): Promise<void> => {
    setResponseState(responseStateValues.loading);
    try {
      const releases = await fetchArtist(inputValue, what);
      setRecList(releases);
      saveQuery(inputValue, what, releases);

      setResponseState(responseStateValues.success);
    } catch (error) {
      setError(error.message);
      setResponseState(responseStateValues.failure);
    }
  };

  const renderSwitch = (responseState: responseStateValues) => {
    switch (responseState) {
      case responseStateValues.loading:
        return (
          <div style={{ width: "90%", textAlign: "center" }}>
            {" "}
            <CircularProgress />{" "}
          </div>
        );
      case responseStateValues.success:
        return <RecordsList list={recList} />;
      case responseStateValues.failure:
        return <ErrorInfo error={error} />;
      default:
        return null;
    }
  };

  return (
    <div>
      <AppBar position="static">
        <Toolbar>
          <Typography variant="h6">Records Search App</Typography>
        </Toolbar>
      </AppBar>
      <Box width="90%" m={4}>
        <Input
          onChange={handleInputChange}
          type="search"
          id="search"
          name="searchByArtist"
          data-value={inputValue}
        />
        <Button
          variant="contained"
          className={classes.button}
          color="primary"
          onClick={() => saveSearchResult("artist")}
        >
          By artist
        </Button>
        <Button
          variant="contained"
          className={classes.button}
          color="primary"
          onClick={() => saveSearchResult("label")}
        >
          By label
        </Button>
        <Box m={4}>{renderSwitch(responseState)}</Box>
        {queryHistory.length && (
          <QueryHistory
            clear={clearHistory}
            history={queryHistory}
            update={deleteHistoryItem}
          />
        )}
      </Box>
      <AppBar position="static" className={classes.toolbar}>
        <Toolbar>
          <Typography variant="subtitle2">Puchweb 2020</Typography>
        </Toolbar>
      </AppBar>
    </div>
  );
};

export default App;
