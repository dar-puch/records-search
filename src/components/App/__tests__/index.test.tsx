import * as React from "react";
import { shallow, mount } from "enzyme";
import "jest-styled-components";

import App from "../index";

describe("<App/>", () => {
  xit("renders correctly", () => {
    expect(shallow(<App />)).toMatchSnapshot();
  });

  xit("should have proper data-value attribute when input is altered", () => {
    const event = {
      target: { value: "something" }
    };
    const wrapper = shallow(<App />);
    wrapper.find("#search").simulate("change", event);
    wrapper.update();
    expect(wrapper.find("#search").prop("data-value")).toEqual("something");
  });

  xit("should call saveSearchResult with given parameter when button is clicked", () => {
    const saveSearchResultMock = jest.fn();
    const component = shallow(<App />);
    component.find("button").simulate("click");
    expect(saveSearchResultMock).toBeCalledWith("artist");
  });

  xit("should render a RecordList when response is success", () => {
    const recList = [
      {
        id: 1,
        title: "Jason",
        year: "1989",
        catno: "ff45",
        cover_image: "hhh"
      },
      { id: 2, title: "Olion", year: "1955", catno: "787r", cover_image: "hhh" }
    ];
  });
});
