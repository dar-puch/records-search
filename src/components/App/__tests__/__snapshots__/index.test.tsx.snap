// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> renders correctly 1`] = `
<div>
  <WithStyles(ForwardRef(AppBar))
    position="static"
  >
    <WithStyles(ForwardRef(Toolbar))>
      <WithStyles(ForwardRef(Typography))
        variant="h6"
      >
        Discon
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Toolbar))>
  </WithStyles(ForwardRef(AppBar))>
  <Styled(MuiBox)
    m={4}
    width="90%"
  >
    <WithStyles(ForwardRef(Input))
      data-value=""
      id="search"
      name="searchByArtist"
      onChange={[Function]}
      type="search"
    />
    <WithStyles(ForwardRef(Button))
      className="makeStyles-button-2"
      color="primary"
      onClick={[Function]}
      variant="contained"
    >
      By artist
    </WithStyles(ForwardRef(Button))>
    <WithStyles(ForwardRef(Button))
      className="makeStyles-button-2"
      color="primary"
      onClick={[Function]}
      variant="contained"
    >
      By label
    </WithStyles(ForwardRef(Button))>
    <Styled(MuiBox)
      m={4}
    />
    0
  </Styled(MuiBox)>
  <WithStyles(ForwardRef(AppBar))
    className="makeStyles-toolbar-1"
    position="static"
  >
    <WithStyles(ForwardRef(Toolbar))>
      <WithStyles(ForwardRef(Typography))
        variant="subtitle2"
      >
        Puchweb 2020
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Toolbar))>
  </WithStyles(ForwardRef(AppBar))>
</div>
`;

exports[`<App/> should render a RecordList when response is success 1`] = `
<div>
  <WithStyles(ForwardRef(AppBar))
    position="static"
  >
    <WithStyles(ForwardRef(Toolbar))>
      <WithStyles(ForwardRef(Typography))
        variant="h6"
      >
        Discon
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Toolbar))>
  </WithStyles(ForwardRef(AppBar))>
  <Styled(MuiBox)
    m={4}
    width="90%"
  >
    <WithStyles(ForwardRef(Input))
      data-value=""
      id="search"
      name="searchByArtist"
      onChange={[Function]}
      type="search"
    />
    <WithStyles(ForwardRef(Button))
      className="makeStyles-button-2"
      color="primary"
      onClick={[Function]}
      variant="contained"
    >
      By artist
    </WithStyles(ForwardRef(Button))>
    <WithStyles(ForwardRef(Button))
      className="makeStyles-button-2"
      color="primary"
      onClick={[Function]}
      variant="contained"
    >
      By label
    </WithStyles(ForwardRef(Button))>
    <Styled(MuiBox)
      m={4}
    />
    0
  </Styled(MuiBox)>
  <WithStyles(ForwardRef(AppBar))
    className="makeStyles-toolbar-1"
    position="static"
  >
    <WithStyles(ForwardRef(Toolbar))>
      <WithStyles(ForwardRef(Typography))
        variant="subtitle2"
      >
        Puchweb 2020
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Toolbar))>
  </WithStyles(ForwardRef(AppBar))>
</div>
`;
